Add scouter java framework

Marcel-Ha opened the pull request:
This adds a first implementation of the scouter agent to the frameworks.
It is unclear to me if it works correctly and if I configured it correctly. For instance should it be MonitoredClassSimple or ...Threaded?

### Important

Depending on the used java version (probably >=11 or >=17) the file `scouter/server/startup.sh` needs to be changed:
`--add-opens java.base/java.lang=ALL-UNNAMED --add-exports java.base/sun.net=ALL-UNNAMED` has to be added to the java command.
The whole scouter folder is downloaded during the benchmark process, so if the file needs to be changed, the benchmark process needs to be aborted after the download and extraction is complete. 
To automatically update the command is an open todo for after I know if it otherwise works correctly.

DaGeRe commented:
Thanks for the PR. Usually, every APM agent should be runnable with Java 11, 17 and 21. Could you add an appropriate GH Workflow (like https://github.com/kieker-monitoring/moobench/blob/main/.github/workflows/executeKiekerJava.yaml) and check this?

You included some snippet from the old Kieker implementation, where --add-opens is added: https://github.com/Marcel-Ha/moobench/blob/10bf04bbedbcd02f7063d2b04ae1f722025cced7/frameworks/Scouter-java/benchmark.sh#L14 could you--if necessary--add this for Scouter?

The monitored class can be omitted: https://github.com/kieker-monitoring/moobench/blob/3a4d9580d223275edb5d6c1af46c68c9655bd927/tools/benchmark/src/main/java/moobench/benchmark/BenchmarkMain.java#L132 So I'd recommend to do it this way. This is code written by developers before me - probably it would make sense to set it to `MonitoredClassThreaded` if threads is set to more than 1. But that would be something to further investigate (whether there is a notable performance difference for 1 thread and the different classes), so let's leave out the parameter for now.

DaGeRe commented:
I've removed the application specification also for Kieker-java: https://github.com/kieker-monitoring/moobench/commit/27303045ad7973325830a5e7ce9b1c11c5c8b15e

DaGeRe commented:
I've checked this again - the default setting is that the application is set to `MonitoredClassSimple`, which is done via the `$MORE_PARAMS` setting: https://github.com/kieker-monitoring/moobench/blob/e05f1c080fc165c98fa1ad51d7d4807b261078b6/frameworks/OpenTelemetry-java/config.rc#L12 
Not doing so leads to increased measured values: https://kieker-monitoring.github.io/moobench/dev/bench/ I do not know why there is the `MonitoredClassSimple` and `MonitoredClassThreaded` exist (unfortunately, its not documented in the javadoc...). But that would be something for further investigation, for now, I'd stick with the current design.

Marcel-Ha commented:
I added that parameter now and also an APP_CLASS parameter to reference the class for one of the parameters for scouter (hook_service_patterns).

I also removed the Java hotfix for kieker. The scouter server sadly does not use the JAVA_OPTS but I tested it: If the add-opens/-exports parameter are not added in the startup.sh of the Scouter server there is an exception but the server still starts. So we can maybe just ignore it?

I am also wondering what exactly we want to benchmark. The agent seems to work without the server. So if its just the overhead of the agent the whole server setup could maybe be skipped.

DaGeRe commented:
Thanks for the update. Could you work on another branch (support-scouter) instead of main? That would make it easier to try out your changes, then I can have a look at the `JAVA_OPTS` situation.

Regarding benchmarking: We benchmark the whole process, from obtaining the data to serializing them. We assume that if there are two processes on the same machine, that the second process (for example a span receiving Zipkin) doesn't influence the MooBench application too much (which might not be true in all cases, but it gives us an estimate of the overhead). For scouter, we therefore also need some data sink that processes the data (and might discard them afterwards).

Marcel-Ha commented:
New Pull request with different source branch and changes from main merged in: https://github.com/kieker-monitoring/moobench/pull/22
